public/pages/workflow_detail/workspace/workspace.tsx

```diff
@@ -3,24 +3,161 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-import React from 'react';
+import React, { useRef, useContext, useCallback, useEffect } from 'react';
+import ReactFlow, {
+  Controls,
+  Background,
+  useNodesState,
+  useEdgesState,
+  addEdge,
+} from 'reactflow';
 import { useSelector } from 'react-redux';
-import { EuiFlexGroup, EuiFlexItem } from '@elastic/eui';
-import { AppState } from '../../../store';
-import { WorkspaceComponent } from '../workspace_component';
+import { EuiFlexItem, EuiFlexGroup } from '@elastic/eui';
+import { AppState, rfContext } from '../../../store';
+import { convertToReactFlowData } from '../../../../common';
 
-export function Workspace() {
-  const { components } = useSelector((state: AppState) => state.workspace);
+// styling
+import 'reactflow/dist/style.css';
+import './reactflow-styles.scss';
+
+// eslint-disable-next-line @typescript-eslint/no-empty-interface
+interface WorkspaceProps {}
+
+export function Workspace(props: WorkspaceProps) {
+  const reactFlowWrapper = useRef(null);
+  const { reactFlowInstance, setReactFlowInstance } = useContext(rfContext);
+
+  // Fetching workspace state to populate the initial nodes/edges.
+  // Where/how the low-level ReactFlow JSON will be persisted is TBD.
+  // TODO: update when that design is finalized
+  const storedComponents = useSelector(
+    (state: AppState) => state.workspace.components
+  );
+  const { rfNodes, rfEdges } = convertToReactFlowData(storedComponents);
+  const [nodes, setNodes, onNodesChange] = useNodesState(rfNodes);
+  const [edges, setEdges, onEdgesChange] = useEdgesState(rfEdges);
+
+  const onConnect = useCallback(
+    (params) => {
+      setEdges((eds) => addEdge(params, eds));
+    },
+    // TODO: add customized logic to prevent connections based on the node's
+    // allowed inputs. If allowed, update that node state as well with the added
+    // connection details.
+    [setEdges]
+  );
+
+  const onDragOver = useCallback((event) => {
+    event.preventDefault();
+    event.dataTransfer.dropEffect = 'move';
+  }, []);
+
+  const onDrop = useCallback(
+    (event) => {
+      event.preventDefault();
+      // Get the node info from the event metadata
+      const nodeData = event.dataTransfer.getData('application/reactflow');
+
+      // check if the dropped element is valid
+      if (typeof nodeData === 'undefined' || !nodeData) {
+        return;
+      }
+
+      // Fetch bounds based on the ref'd div component, adjust as needed.
+      // TODO: remove hardcoded bounds and fetch from a constant somewhere
+      // @ts-ignore
+      const reactFlowBounds = reactFlowWrapper.current.getBoundingClientRect();
+      // @ts-ignore
+      const position = reactFlowInstance.project({
```

Comment:
just wondering, where does the height and width info come from for these nodes to be rendered?

Reply:
The `reactFlowWrapper` contains the element positioning in the webpage, and is used for calculating the bounds on line 69. The `position` based on these hardcoded bounds is saying that for every element that is dragged into the workspace, place it 80 pixels from the left bounds, and 90 pixels from the top bounds. We will likely want this to be dynamic based on the user mouse, for example, but just placing those hardcoded values for ease of use for now.